### PR TITLE
Refactor, fix Polymer 2.x race condition

### PR DESCRIPTION
**Pull request by davidrleonard**
This refactor keeps the template and API the same but cleans up
the code a bit and moves the responsibility for showing/hiding
the deck selector dropdown to a single private property. This
property could be exposed publicly if we want to allow the
developer to set it.

This also fixes an issue that appeared in Polymer 2 where the
dropdown would toggle closed then immediately toggle opened
when the user tapped a new deck to select it.

@PredixDev/px anyone up for a review?

### CODE REVIEW

File: px-deck-selector.html [hunk 1]
```diff
@@ -33,18 +33,18 @@
...
 
-    <div class$="dropdown-text actionable actionable--action {{_hideChevronIfNoDecks(_isDecks)}}" on-click="_toggleDropdown">
+    <div id="trigger" class$="dropdown-text actionable actionable--action [[_hideChevronIfNoDecks(decks)]]" on-click="_handleTriggerTapped">
       <span class="truncate" title={{selectedDeck.name}}>{{selectedDeck.name}}</span>
       <px-icon icon="px-utl:chevron"></px-icon>
     </div>
-    <ul id="dropdown" class="dropdown-menu list-bare visuallyhidden">
+    <ul id="dropdown" class$="dropdown-menu list-bare [[_getDropdownClassName(_isDropdownOpen)]]">
       <template is="dom-repeat" items="{{decks}}" as="deck">
-        <li on-click="_selectDeck" class$="dropdown-menu__item {{_isSelected(deck,selectedDeck)}}">
```
Comment (benoitjchevalier) [inline px-deck-selector.html]: should we use on-tap rather than on-click?

File: px-deck-selector.html [hunk 2]
```diff
@@ -122,98 +121,79 @@
...
+      if (Array.isArray(decks) && decks.length > 0) {
+        this.selectedDeck = decks[0];
+      }
     },
 
-    _hideChevronIfNoDecks: function(isDecks){
-      return isDecks ? '' : 'visuallyhidden';
+    _handleDropdownOpenedChanged: function(opened) {
+      if (opened) {
+        document.addEventListener('click', this._handleDocumentClicked, false);
```
Comment (benoitjchevalier) [inline px-deck-selector.html]: tap instead of click?

File: px-deck-selector.html [hunk 3]
```diff
@@ -122,98 +121,79 @@
...
-      return deck.name === selectedDeck.name ? 'selected' : '';
+    /**
+     * When the user taps somewhere else on the page when the dropdown is opened,
+     * close the dropdown. If the user taps on the trigger do nothing and
+     * let the trigger handler close the dropdown.
+     */
+    _handleDocumentTapped: function(e) {
+      const path = Polymer.dom(e).path;
+      const trigger = this.$.trigger;
+      const isTriggerInPath = path.filter(el => el === trigger).length > -1;
```
Comment (benoitjchevalier) [inline px-deck-selector.html]: should it be >0?

File: px-deck-selector.html (hunk 1 above)
Comment (davidrleonard) [reply px-deck-selector.html]: 👍

File: px-deck-selector.html (hunk 2 above)
Comment (davidrleonard) [reply px-deck-selector.html]: I don't think there is a native `tap` event in the browser, i think `on-tap` is just Polymer sugar. I could be wrong though.

File: px-deck-selector.html (hunk 3 above)
Comment (davidrleonard) [reply px-deck-selector.html]: 👍  good catch

**Pull request merged**